### PR TITLE
Tier-0 generic execution Jenkins Job configuration

### PR DESCRIPTION
**Pull request by sunilkumarn417**
- Added Tier-0 generic Jenkins configuration Groovy script
- Created Job already in Jenkins Server.

https://ceph-downstream-jenkins-csb-storage.apps.ocp4.prod.psi.redhat.com/job/rhceph-generic-tier-0/17/console

Signed-off-by: sunilkumarn417 <sunnagar@redhat.com>

### CODE REVIEW

**Comment by sunilkumarn417** [comment]
**latest results:** https://ceph-downstream-jenkins-csb-storage.apps.ocp4.prod.psi.redhat.com/job/rhceph-generic-tier-0/38/console

**Comment by sunilkumarn417** [comment]
https://ceph-downstream-jenkins-csb-storage.apps.ocp4.prod.psi.redhat.com/job/rhceph-generic-tier-0/39/console

File: pipeline/executors/Jenkinsfile-tier-0.groovy [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+                "type": "tier0",
+                "result": currentBuild.currentResult,
+                "url": env.BUILD_URL,
+                "log": "${env.BUILD_URL}/console",
+            ]
+        ]
+
+        def msgContent = writeJSON returnText: true, json: artifactsMap
+        println "${msgContent}"
+
```
Comment (psathyan) [inline pipeline/executors/Jenkinsfile-tier-0.groovy:113]: Let's discuss with Chaithra what is needed from the library.

File: pipeline/executors/Jenkinsfile-tier-0.groovy [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+                submoduleCfg: [],
+                userRemoteConfigs: [[
+                    url: 'https://github.com/red-hat-storage/cephci.git'
+                ]]
+            ])
+
+            // prepare the node for executing test suites
+            sharedLib = load("${env.WORKSPACE}/pipeline/vars/lib.groovy")
+            sharedCommon = load("${env.WORKSPACE}/pipeline/vars/common.groovy")
+            sharedCommon.prepareNode()
```
[the remark this answers is not in the document]
Comment (sunilkumarn417) [reply pipeline/executors/Jenkinsfile-tier-0.groovy]: done.

File: pipeline/vars/common.groovy [hunk 3]
```diff
@@ -473,6 +484,18 @@ def getRHCSVersion() {
...
+    /*
+        Returns the RHCEPH version from the compose ID in CI_MESSAGE.
+    */
+    def compose = getCIMessageMap()
+    def ver
+    if ( getCvpVariable() ) {
+        ver = compose.artifact.nvr
+    }
+    return ver
+}
```
[the remark this answers is not in the document]
Comment (sunilkumarn417) [reply pipeline/vars/common.groovy:497]: deleted.

File: pipeline/vars/lib.groovy [hunk 4]
```diff
@@ -215,4 +224,60 @@ def SendUMBMessage(def msgMap, def overrideTopic, def msgType){
...
+    catchError (message: 'STAGE_FAILED', buildResult: 'FAILURE', stageResult: 'FAILURE') {
+        try {
+            sh(script: "sh ${scriptPath} ${cliArgs}")
+        } catch(Exception err) {
+            rc = 1
+            println err.getMessage()
+            error "Encountered an error"
+        } finally {
+            println "finally..... completed!!!"
+        }
```
[the remark this answers is not in the document]
Comment (sunilkumarn417) [reply pipeline/vars/lib.groovy:241]: Deleted.

File: pipeline/vars/lib.groovy [hunk 5]
```diff
@@ -215,4 +224,60 @@ def SendUMBMessage(def msgMap, def overrideTopic, def msgType){
 
 }
 
+def executeTestScript(def scriptPath, def cliArgs) {
+   /*
+        Executes the test script
+   */
+    def rc = 0
```
[the remark this answers is not in the document]
Comment (sunilkumarn417) [reply pipeline/vars/lib.groovy:231]: Done.

File: pipeline/vars/lib.groovy [hunk 6]
```diff
@@ -215,4 +224,60 @@ def SendUMBMessage(def msgMap, def overrideTopic, def msgType){
...
+
+        MAJOR   -   RHceph major version (ex., 5)
+        MINOR   -   RHceph minor version (ex., 0)
+        TIER-x  -   Tier level number (ex., tier-0)
+    */
+    def RHCSVersion = getRHCSVersionFromArtifactsNvr()
+    def majorVersion = RHCSVersion["major_version"]
+    def minorVersion = RHCSVersion["minor_version"]
+
+    def scriptPath = "./pipeline/scripts/${majorVersion}/${minorVersion}/${tierLevel}/"
```
[the remark this answers is not in the document]
Comment (sunilkumarn417) [reply pipeline/vars/lib.groovy:263]: Done.

File: pipeline/executors/Jenkinsfile-tier-0.groovy [hunk 7]
```diff
@@ -0,0 +1,116 @@
...
+        /* Publish results through E-mail and Google Chat */
+        def emailTo = "ceph-qe@redhat.com"
+        withEnv([
+            "rhcephVersion=${ciMap["artifact"]["nvr"]}",
+            "composeId=${ciMap["artifact"]["version"]}",
+            "composeUrl=${ciMap["build"]["compose-url"]}",
+            "repository=${ciMap["artifact"]["build_action"]}"
+        ]) {
+
+            if ( ! (1 in testResults.values()) ) {
```
[the remark this answers is not in the document]
Comment (sunilkumarn417) [reply pipeline/executors/Jenkinsfile-tier-0.groovy:64]: Done.

File: pipeline/executors/Jenkinsfile-tier-0.groovy (hunk 1 above)
Comment (sunilkumarn417) [reply pipeline/executors/Jenkinsfile-tier-0.groovy:113]: Sure. 
IMO, It would be good to send it from test groovy only.

File: pipeline/executors/Jenkinsfile-tier-0.groovy [hunk 8]
```diff
@@ -0,0 +1,116 @@
...
+
+    stage('Publish Results') {
+        /* Publish results through E-mail and Google Chat */
+        def emailTo = "ceph-qe@redhat.com"
+        withEnv([
+            "rhcephVersion=${ciMap["artifact"]["nvr"]}",
+            "composeId=${ciMap["artifact"]["version"]}",
+            "composeUrl=${ciMap["build"]["compose-url"]}",
+            "repository=${ciMap["artifact"]["build_action"]}"
+        ]) {
```
[the remark this answers is not in the document]
Comment (sunilkumarn417) [reply pipeline/executors/Jenkinsfile-tier-0.groovy:62]: we can check this after portability.

File: pipeline/vars/common.groovy [hunk 9]
```diff
@@ -153,6 +153,17 @@ def executeTest(def cmd, def instanceName) {
...
+    /*
+        Return random Id
+    */
+    return sh(
+        script: "cat /dev/urandom | tr -cd 'a-f0-9' | head -c 5",
+        returnStdout: true
+    ).trim()
+
+}
+
```
Comment (psathyan) [inline pipeline/vars/common.groovy:166]: Can we undo this change?

**Review by psathyan — COMMENTED** [review]
Thanks... minor changes. The rest looks good.

File: pipeline/executors/Jenkinsfile-tier-0.groovy [hunk 10]
```diff
@@ -0,0 +1,127 @@
...
+        if ( ! (sharedLib.failStatus in testResults.values()) ) {
+            emailTo = "ceph-qe-list@redhat.com"
+            releaseContent = sharedLib.ReadFromReleaseFile(majorVersion, minorVersion)
+            releaseContent[tierLevel]["composes"] = releaseContent[buildPhase]["composes"]
+            releaseContent[tierLevel]["last-run"] = releaseContent[buildPhase]["ceph-version"]
+            sharedLib.WriteToReleaseFile(majorVersion, minorVersion, releaseContent)
+        }
+
+        sharedLib.sendGChatNotification(testResults, tierLevel)
+        sharedLib.sendEMail(testResults, artifactDetails, tierLevel)
```
[the remark this answers is not in the document]
Comment (sunilkumarn417) [reply pipeline/executors/Jenkinsfile-tier-0.groovy]: Created new definition which returns artifacts details.

File: pipeline/executors/Jenkinsfile-tier-0.groovy [hunk 11]
```diff
@@ -0,0 +1,127 @@
...
+        def (majorVersion, minorVersion) = getRHCSVersionFromArtifactsNvr()
+
+        /* 
+           Read the release yaml contents to get contents,
+           before other listener/Executo Jobs updates it.
+        */
+        releaseContent = sharedLib.ReadFromReleaseFile(majorVersion, minorVersion, lockFlag=false)
+        def artifactDetails = [
+            "composes": releaseContent[buildPhase]["composes"],
+            "product": "Redhat",
```
[the remark this answers is not in the document]
Comment (sunilkumarn417) [reply pipeline/executors/Jenkinsfile-tier-0.groovy]: Done

File: pipeline/vars/common.groovy (hunk 9 above)
Comment (sunilkumarn417) [reply pipeline/vars/common.groovy]: Done.

File: pipeline/executors/Jenkinsfile-tier-0.groovy [hunk 12]
```diff
@@ -0,0 +1,127 @@
...
+            ],
+            "contact": [
+                "name": "Downstream Ceph QE",
+                "email": "ceph-qe@redhat.com",
+            ],
+            "pipeline": [
+                "name": "rhceph-tier-0",
+            ],
+            "test-run": [
+                "type": "tier0",
```
[the remark this answers is not in the document]
Comment (sunilkumarn417) [reply pipeline/executors/Jenkinsfile-tier-0.groovy]: Done.

**Review by psathyan — APPROVED** [review]
